src/Containers/container.jl

```diff
@@ -113,13 +113,43 @@ function container(f::Function, indices, ::Type{SparseAxisArray})
     mappings = Base.Generator(I -> I => f(I...), indices)
     # Same as `Dict(mapping)` but it will error if two indices are the same.
     data = NoDuplicateDict(mappings)
-    dict = data.dict
-    if isempty(dict)
-        # If `dict` is empty, it was not able to determine the type
-        # of the key hence the type of `dict` is `Dict{Any, Any}`.
-        # This is an issue since `SparseAxisArray` needs the key
-        # type to be a tuple.
-        dict = Dict{_eltype_or_any(indices),Any}()
-    end
-    return SparseAxisArray(dict)
+    return _sparseaxisarray(data.dict, f, indices)
+end
+
+# The NoDuplicateDict was able to infer the element type.
+_sparseaxisarray(dict::Dict, ::Any, ::Any) = SparseAxisArray(dict)
+
+function _sparseaxisarray(dict::Dict{Any,Any}, ::Any, indices)
+    @assert isempty(dict)
+    return SparseAxisArray(Dict{_eltype_or_any(indices),Any}())
+end
+
+# If the eltype is a tuple, we can use that as the key for the SparseAxisArray.
+function _container_dict(
+    K::Type{<:NTuple{N,Any}},
+    f,
+    ::Type{<:NTuple{N,Any}},
+) where {N}
+    ret = Base.return_types(f, K)
+    V = length(ret) == 1 ? first(ret) : Any
+    return Dict{K,V}()
+end
+# Catch this case where Union{} <: Tuple.
+function _container_dict(
+    ::Type{Union{}},
+    ::Any,
+    K::Type{<:NTuple{N,Any}},
+) where {N}
+    return Dict{K,Any}()
+end
+_container_dict(::Any, ::Any, K) = Dict{K,Any}()
+
+# The NoDuplicateDict was not able to infer the element type. To make a
+# best-guess attempt, collect all of the keys excluding the conditional
+# statement (these must be defined, because the conditional applies to the
+# lowest-level of the index loops), then get the eltype of the result.
+function _sparseaxisarray(dict::Dict{Any,Any}, f, indices::NestedIterator)
+    @assert isempty(dict)
+    el = eltype(collect(nested(indices.iterators...)))
```

Comment:
Why not `el = typeof(first(nested(indices.iterators...))))` ?
Instead of using `return_types` (which is discouraged), we could do `eltype(f(first(nested(indices.iterators...))))`

Reply:
Because the elements are not guaranteed to have the same type.

Reply:
I think that's the advantage of `Base.@default_eltype`, see https://github.com/jump-dev/JuMP.jl/issues/2552#issuecomment-814717126
They call `return_types` on `_iterator_upper_bound` with the `Generator` directly.